Enable WebRTC

istankovic opened the pull request:
This PR adds the metadata querying during WebRTC signaling which was the last remaining thing needed to enable WebRTC.

Commit:
raiden: transport: fetch metadata for WebRTC signaling

Also enable WebRTC.

Commit:
tests: adjust to the changes in WebRTCManager

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -22,19 +22,33 @@
 
 class _CoroutineHandler:
     def __init__(self) -> None:
+        super().__init__()
```
Why is this necessary?

istankovic replied:
Because `WebRTCManager` is derived from both `_CoroutineHandler` and `Runnable` and we need to call all the constructors.

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -620,15 +604,17 @@ def stop(self) -> None:
         self._send_hangup_messages()
 
+    def _send_signaling_message(self, address: Address, message: str) -> None:
+        metadata = _query_metadata(self._pfs_config, address)
```
Is the private `_query_metadata` really necessary?

istankovic replied:
Perhaps not as it is, but the idea is to add metadata caching for the duration of the signaling process, and then it will be nicer to have this factored out.

fredo reviewed (approved):
LGTM. Left some comments below. 

Additionally, in web_rtc.py in line 428 this needs to be changed to `via matrix to_device message` as room usage is outdated

Commit:
tests: start a dummy PFS when doing a smoketest

Also switch to PFS routing mode when doing a smoketest.

Commit:
tests: utils: add a way to create dummy PFSConfig

Commit:
tests: remove unneeded type

Commit:
tests: remove query_address_metadata_mock

It's not usable because we need to provide valid metadata,
which includes correctly signed user IDs.

Commit:
raiden: network: introduce PFSProxy

Instead of calling query_address_metadata and query_paths directly,
we should now call methods on PFSProxy. This allows us to tweak the
behaviour and provide valid metadata in tests more easily.

Commit:
raiden: use PFS proxies to query address metadata and paths

Commit:
raiden: transport: remove erroneous check for the raiden service

RaidenService is a Runnable, which implements __bool__ that checks
whether the greenlet is running, which is not what we want to test here.
Instead, we'd like to check against None, but since the service will
not be None here, remove the check completely.

Commit:
raiden: transport: make missing metadata an error

Commit:
tests: fix typo

Commit:
raiden: transport: explicitly check for None

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -620,15 +603,17 @@ def stop(self) -> None:
         self._reset_state()
         self._send_hangup_messages()
```
the second line will be a noop because `reset_state()` will empty the dictionary

istankovic replied:
Indeed, you're right. What was the purpose of that before (my changes didn't touch that part)? It seems like those hangup messages were never sent. Shall I just remove that method?

fredo replied:
sounds reasonable. I think they should be sent as this is called in the stop method to notify the partners that the node is currently shutting down the channels
(In the past we had a failure in state showing that the datachannel was open although the partner already closed it)

fredo replied:
Maybe calling the hangup method before closing the channels makes sense

istankovic replied:
> Maybe calling the hangup method before closing the channels makes sense

Yes, I'll do that and see if it goes well.

fredo commented on raiden/network/transport/matrix/transport.py:
```diff
@@ -579,8 +583,7 @@ def stop(self) -> None:
 
-        if self._raiden_service:
-            self._web_rtc_manager.stop()
+        self._web_rtc_manager.stop()
```
I think the condition was there for a reason (although it was a false condition). Now, with the change that the web_rtc_manager is only initialized in `start()`, the argument for the if condition is even stronger.

If `stop()` is called before `start()` then `self.web_rtc_manager` will be none. I wonder why the linter does not prevent that.

istankovic replied:
If `stop` is called before `start`, it will exit immediately because of the `_stop_event` check. So it will all be good. In fact, I'm going to explicitly reset `_web_rtc_manager` to `None` to make it even more clear that it's valid only in the time period `<start, stop>`.

fredo reviewed (changes requested):
LGTM in general, but please have a look again if removing the if condition was the right thing to do.

Commit:
raiden: transport: send hangup messages before closing the connections

Commit:
raiden: transport: add a way to ask whether the transport has started

Commit:
raiden: raiden_service: stop the alarm before transport

Commit:
raiden: raiden_service: don't do health check if the transport hasn't started yet
